ViewModel 분리 작업 진행 중

june0122 opened the pull request.

Commit:
MVVM 자료

june0122 committed:
feat: ViewModel로부터 받은 메시지를 뷰에 표시하기 위한 Event 클래스 정의

june0122 committed:
refactor: Toast 메시지 확장 함수 정의

june0122 committed:
refactor: 파라미터가 있는 ViewModel 객체 생성을 위해 Factory 클래스 구현

ViewModelProvider.Factory 인터페이스를 구현한 Factory 클래스 정의

june0122 committed:
refactor: 예제 ViewModel 클래스 삭제

june0122 committed:
refactor: PlantListViewModel 패키지 분리

june0122 committed:
refactor: 뷰 컨트롤러에 할당된 과도한 책임을 ViewModel로 분리

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -30,72 +26,85 @@ class PlantListFragment : Fragment() {
-    private var lastPage = 0
+    private val viewModel: PlantListViewModel by viewModels(
+        factoryProducer = { PlantListViewModelFactory(plantListAdapter) }
+    )
```
👍

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -30,72 +26,85 @@ class PlantListFragment : Fragment() {
-                    .addToBackStack(null)
-                    .commit()
+                viewModel.onPlantClick(position)
+                viewModel.showDetail.observe(viewLifecycleOwner) { plantData ->
```
observe 코드는 onClick 일어날때마다 할 필요성이 없으니, 여기에 위치하면 안됩니다.

여기에 위치했을때 어떠한 일이 발생하는지 직접 확인해보세요. 로그를 출력해보시거나, (화면 이동을 주석으로 달고요)

june0122 replied:
Observer가 LiveData 객체를 구독해서 어짜피 변경사항에 대한 알림을 받게 되어있는데 생각이 짧았네요🥲

observe 코드를 밖으로 빼니 DetailFragment(상세 화면)에서 뒤로 가기를 할 경우, observe 코드가 다시 변경을 감지하고 코드를 실행시켜서 상세 화면을 빠져나오지 못하는 버그가 있는데 [이 내용](https://stackoverflow.com/questions/58342060/why-is-onchanged-being-called-when-i-navigate-back-to-a-fragment)인 것 같아 참고해서 최종적으로 수정하겠습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -30,72 +26,85 @@ class PlantListFragment : Fragment() {
-                PlantDialogFragment.newInstance(plantData).apply {
-                    show(this@PlantListFragment.parentFragmentManager, DIALOG_PLANT)
+                viewModel.onPlantClick(position)
+                viewModel.showDetail.observe(viewLifecycleOwner) { plantData ->
```
여기도 위와 동일

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -30,72 +26,85 @@ class PlantListFragment : Fragment() {
+        plantRecyclerView.adapter = plantListAdapter
+        plantRecyclerView.layoutManager = layoutManager
+        plantRecyclerView.addItemDecoration(PlantListItemDecoration(viewModel.spanCount, 60, true))
+        plantRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() { // < ---- 액션의 등록 및 시작
```
이런 경우엔 binding.rvPlantList.run {}으로 묶어주는게 깔끔하겠죠.

```
plantRecyclerView = binding.rvPlantList
plantRecyclerView.adapter = plantListAdapter
plantRecyclerView.layoutManager = layoutManager
plantRecyclerView.addItemDecoration(PlantListItemDecoration(viewModel.spanCount, 60, true))
plantRecyclerView.addOnScrollListener(
```

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -30,72 +26,85 @@ class PlantListFragment : Fragment() {
+                ) {
+                    recyclerView.post { viewModel.deleteProgress() }
+                    viewModel.currentPage++
+                    if (viewModel.currentPage <= viewModel.lastPage) viewModel.getUserList()
```
77~85 라인 부분을 작성해주셨던 시나리오 형태로 변경해보세요.
이렇게 작성되면 기존 보다 더 깔끔하지 않고,

특히 viewModel.getValue() 형태를 사용하시면 ViewModel을 사용하는 의미가 사라집니다.

VieModel에서는 값을 통지해주는것이지, 값을 직접 접근해서 가져가는 형태이면 안됩니다.
View -> viewModel 함수를 호출(이벤트 전달) -> ViewModel이 모델에 데이터를 요청하거나, ViewModel 스스로 데이터를 처리 -> ViewModel 내부에 있는 옵저버(LiveData, Flow, RxJava 등)에게 값을 통지해주고 끝나야해요.

currentPage, itemCount 역시 직접 접근해서 사용할 수 없도록 되어야 하는 형태로 private으로 정의하시고 다시 고민해보세요.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -30,72 +26,85 @@ class PlantListFragment : Fragment() {
 
+    private fun canLoaded(canScrollVertically: Boolean, lastVisibleItemPosition: Int, totalCount: Int): Boolean =
+        canScrollVertically && lastVisibleItemPosition == totalCount
+
```
이 함수 역시 ViewModel로 이동 시켜보세요.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/utils/Event.kt:
```diff
@@ -0,0 +1,24 @@
+     * Returns the content, even if it's already been handled.
+     */
+    fun peekContent(): T = content
+}
```
이 이벤트를 추가한 이유가 뭘까요?

june0122 replied:
(밑에 nullable로 처리한 Context 포함) ViewModel 내부에서 Toast로 띄울 메시지를 전달하기 위해 방법들을 실험해보다가 함께 push 해버렸는데 이 부분은 좀 더 연구해보겠습니다! 

참고 자료: https://stackoverflow.com/a/62113974/12364882

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/utils/Extensions.kt:
```diff
@@ -19,4 +20,7 @@ fun convertDpToPixel(context: Context, dp: Int): Float {
+}
+
+fun Context?.toast(text: String, duration: Int = Toast.LENGTH_SHORT) =
+    this?.let { Toast.makeText(it, text, duration).show() }
```
context를 널로 한 이유가 뭘까요?

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListPresenter.kt:
```diff
@@ -0,0 +1,21 @@
+        //
+        view.fail()
+    }
+}
```
아마 이 코드는 저와 이야기하면서 추가한 코드로 보이는데, 여기서는 필요치 않아보입니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+    var currentPage = 1
+    var perPage = 20
+    var lastPage = 0
+    var spanCount = 2
```
이 3개의 값은 private으로 처리하시고, spanCount는 ViewModel이 알필요 없어보이네요.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+
+    private val statusMessage = MutableLiveData<Event<String>>()
+    val message: LiveData<Event<String>>
+        get() = statusMessage
```
외부에서도 보는 메시지는 statusMessage일것 같은데, 네이밍 규칙을 잘 고민해보세요.

june0122 replied:
테스트용으로 선언한 변수들인데 변수명을 꼼꼼히 생각하지 못했네요. 잘 확인하겠습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+class PlantListViewModel(
+    private val plantListAdapter: PlantListAdapter,
+) : ViewModel(), PlantClickListener {
+    val items = plantListAdapter.items
```
어댑터의 아이템을 다시 items으로 맵핑할 필요가 있을까요?

june0122 replied:
해당 클래스에서 `plantListAdapter.items`에 접근할 일이 꽤 있어서 `items`로 맵핑하면 간단하게 사용할 수있지 않을까? 라는 생각으로 작성한 코드인데 변수명의 중복이 문제가 되는 걸까요? 코드를 보다니 헷갈릴 수 있는 여지가 있는 것 같기도 하네요.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+    fun deleteProgress() {
+        if (items.last().description == STATUS_LOADING) {
+            items.removeAt(items.lastIndex)
+            plantListAdapter.notifyItemRemoved(items.lastIndex + 1)
```
이 함수는 ViewModel 내부에서만 사용하도록하고, 위에서 작성한 scroll에서 더 불러오기를 뷰모델 내부에서 처리한다면 좀 더 깔끔해질거에요. 나중에 변경해보세요.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    fun updateUserList(users: Users) {
+        lastPage = (users.total_count / perPage) + 1
```
lastPage를 미리 +1 했을때 어떤 문제가 있어보입니다.
- 네트워크 통신에 실패한다면?
- 데이터를 가져왔는데 실제 lastPage와 내가 알던 lastPage가 달라졌다면?

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+        })
+    }
+
+    fun updateUserList(users: Users) {
```
함수 접근제어자가 필요하지 않을까요?

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+        users.items.forEach {
+            items.add(
+                Plant(imageUrl = it.avatarUrl, name = it.login, description = "")
+            )
```
데이터 무결성이 깨질것 같지 않을까요?
사용자가 스크롤하는 도중에도 items에 값이 추가되어지고 있으니깐요?

이 경우 forEach 대신 map을 활용하고, 새로운 리스트 아이템을 업데이트하도록 변경해보세요.

june0122 replied:
```kotlin
val newData = users.items.map {
            Plant(imageUrl = it.avatarUrl, name = it.login, description = "")
        }
items.addAll(newData)
plantListAdapter.notifyItemRangeInserted(itemCount, users.items.size)
```

우선은 맞다고 생각되는 코드로 수정하였습니다!

그런데 `map`과 `forEach`를 사용했을때 데이터 무결성에 왜 차이가 발생하는지 잘 모르겠습니다..

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,91 @@
+
+        if (currentPage < lastPage) {
+            items.add(Plant("", "", STATUS_LOADING)) // progressbar 보여주기 위한 아이템 1개 추가
+            plantListAdapter.notifyItemInserted(itemCount + 1)
```
progress 바가 먼저 보여지는건가요?

올바르게 수정이 필요할것 같은데
- 데이터 불러올때 만 Progress가 필요하다
- 데이터를 다 불러오고나면 Progress가 사라져야 한다.

현재 코드상 데이터가 업데이트하면 currentPage < lastPage?니깐 더 불러올 데이터가 없어도 노출될것 같네요.

june0122 committed:
chore: build.gradle 버전 업데이트

june0122 committed:
refactor: 예제 코드 삭제

june0122 committed:
style: 상세 화면의 background color 추가

june0122 committed:
fix: LiveData에 기존값이 재통보되는 문제를 해결하기 위한 Event Wrapper 추가

june0122 committed:
refactor: 뷰 컨트롤러에 정의된 UI 관련 데이터 분리

june0122 committed:
resource: dimens.xml에 마진값들 추가

june0122 committed:
feat: 로딩 프로세스바에 포커싱될 수 있도록 smoothScroll 기능 추가

june0122 committed:
fix: 프로그래스바가 삭제되면서 좌표값이 이동해 흔들리는 버그 수정

프로그래스바 아이템 삭제 시 position 값이 -1이 되어 프로그래스바가 흔들리는 문제 수정
spanCount가 홀수일 때 여전히 프로그래스바가 흔들리는 문제가 있으므로 수정 필요

june0122 committed:
refactor: 프로그래스바가 데이터 로딩 완료 후에 삭제될 수 있도록 리팩토링

june0122 committed:
fix: 리사이클러뷰 애니메이션을 없애는 것으로 프로그래스바 삭제 시 버그 수정

june0122 committed:
refactor: GithubService의 구현체를 변수를 통해 접근할 수 있도록 변경

동반 객체 내에서 create() 메서드를 통해 생성했던 GithubService의 구현체를
변수를 통해 더욱 간단히 접근할 수 있도록 변경

june0122 committed:
refactor: sealed class를 통해 하위 클래스의 종류를 제한하여 아이템 타입 표시

리사이클러뷰의 아이템 중 프로그래스바가 담긴 아이템을 표현하기 위해 더미 Plant 데이터를
사용하던 코드에서 sealed class인 PlantData를 만들어 이 클래스를 상속하는 Plant,
Progress라는 클래스를 선언. 이로써 이 두 개의 클래스 외에 PlantData 클래스를 상속받는
다른 클래스가 없다는 것을 컴파일러에게 알려준다.

june0122 committed:
refactor: `viewModel.items`를 `items`로 맵핑하지 않고 사용하도록 수정

june0122 committed:
refactor: sealed class `PlantData`를 정의하여 아이템 타입 체크법을 리팩토링

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/api/GithubService.kt:
```diff
@@ -22,23 +22,20 @@ interface GithubService {
+    addConverterFactory(GsonConverterFactory.create())
+}.build()
+
+val githubService: GithubService = retrofit.create(GithubService::class.java)
```
이렇게 분리하였다면, service와 별도의 파일로 만들어두시면 재사용성이 높아질 수 있습니다.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/adapter/PlantListAdapter.kt:
```diff
@@ -6,25 +6,22 @@ import androidx.recyclerview.widget.RecyclerView
 import com.june0122.sunflower.utils.PlantClickListener
 
 const val VIEW_TYPE_ITEM = 0
 const val VIEW_TYPE_LOADING = 1
```
이런 값의 정의는 companion object {}에 정의해두는게 나중에 규모가 커졌을때도 좋아요. 중복 코드 방지도 가능하구요.

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -1,110 +1,123 @@
+
+        binding.rvPlantList.run {
+            adapter = plantListAdapter
+            this.layoutManager = layoutManager
```
run을 사용했을 경우에 발생하는 부분인데 더 좋은 방법은 무엇일까요?
run 대신 다른걸 활용하는 방법으로 고민해보셔도 좋아요.

june0122 replied:
객체 초기화 시 사용하는 `apply`를 이용하는게 더 맞겠네요..!

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/ui/fragment/PlantListFragment.kt:
```diff
@@ -113,46 +126,14 @@ class PlantListFragment : Fragment() {
+    private fun setSpanSize(layoutManager: GridLayoutManager) {
+        layoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
+            override fun getSpanSize(position: Int): Int {
+                return when (viewModel.checkItemType(position)) {
```
viewModel.checkItemType()을 활용하시면 여전히 get의 형태가 남아있는 형태입니다.
아쉽지만 ViewModel을 활용하기로 하였으니 get() 함수는 존재하지 않아야 합니다.

checkItemType()은 ViewModel과는 관련성이 없을수도 있으니 좀 더 고민해보세요.

theo-f-lab replied:
참고로 getItemViewType은 Adapter에 포함되어있고, 구현하셨을거에요. 굳이 ViewModel로 넘겨서 처리할 필요가 없겟죠.
```
mLayoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
            @Override
            public int getSpanSize(int position) {
                switch(mAdapter.getItemViewType(position)){
                    case MyAdapter.TYPE_HEADER:
                        return 2;
                    case MyAdapter.TYPE_ITEM:
                        return 1;
                    default:
                        return -1;
                }
            }
        });
```

theo-f-lab commented on app/src/main/java/com/june0122/sunflower/viewmodel/PlantListViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+
+    fun checkItemType(position: Int): Int {
+        return if (position == itemCount) VIEW_TYPE_LOADING else VIEW_TYPE_ITEM
+    }
```
ViewModel의 함수는 Unit 이어야 합니다.라는 부분을 강조드리면 이런 제약을 가지고 ViewModel의 코드를 작성해보세요.

june0122 committed:
refactor: Service와 Retrofit 인스턴스 생성 코드 분리

june0122 committed:
refactor: 상수값을 동반 객체 내에 정의하기

추후 코드의 규모가 커졌을 때 중복 코드 방지도 되고 좋다.

june0122 committed:
refactor: ViewModel 내부에 리턴 타입이 Unit인 함수로만 구성 및 함수 분리
